lib/src/constants.dart

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/constant/value.dart';
+
+/// Always returns `null` (used as a default for `defaultTo` methods).
+Null _alwaysNull() => null;
+
+/// Returns whether or not [object] is or represents a `null` value.
+bool _isNull(DartObject object) => object?.isNull != false;
+
+/// Similar to [DartObject.getField], but traverses super classes.
+DartObject _getFieldRecursive(DartObject object, String field) {
+  if (_isNull(object)) {
+    return null;
+  }
+  final result = object.getField(field);
+  if (_isNull(result)) {
+    return _getFieldRecursive(object.getField('(super)'), field);
+  }
+  return result;
+}
+
+/// A wrapper for analyzer's [DartObject] with a predictable high-level API.
+///
+/// Unlike [DartObject.getField], all `readX` methods attempt to access super
+/// classes for the field value if not found.
+abstract class Constant {
+  factory Constant(DartObject object) =>
+      _isNull(object) ? const _NullConstant() : new _Constant(object);
+
+  /// Returns whether this constant represents a `bool` literal.
+  ///
+  /// If `true`, [boolValue] will return either `true` or `false` (not throw).
+  bool get isBool;
+
+  /// Returns this constant as a `bool` value.
+  ///
+  /// Throws [FormatException] if [isBool] is `false`.
+  bool get boolValue;
+
+  /// Returns whether this constant represents an `int` literal.
+  ///
+  /// If `true`, [intValue] will return an `int` (not throw).
+  bool get isInt;
+
+  /// Returns this constant as an `int` value.
+  ///
+  /// Throws [FormatException] if [isInt] is `false`.
+  int get intValue;
+
+  /// Returns whether this constant represents a `String` literal.
+  ///
+  /// If `true`, [stringValue] will return a `String` (not throw).
+  bool get isString;
+
+  /// Returns this constant as an `String` value.
+  ///
+  /// Throws [FormatException] if [isString] is `false`.
+  String get stringValue;
+
+  /// Returns whether this constant represents `null`.
+  bool get isNull;
+
+  /// Reads[ field] from the constant as another constant value.
+  Constant read(String field);
+
+  /// Reads [field] from the constant as a boolean.
+  ///
+  /// If the resulting value is `null`, uses [defaultTo] if defined.
+  bool readBool(String field, {bool defaultTo()});
```

Comment:
A DartObject can either _be_ a literal or _have_ fields (that are either literals or objects themselves).

Because `Constant` wraps a `DartObject`, it makes sense to make this easier. It also enables (eventually) added support for using a dot notation, or something like:

```dart
constant.readBool('from.another.thing');
```

If you think it's not worth it and we should just have `xValue`, happy to revert though.